fix: flakiness in org.json.junit.XMLTest#testIndentComplicatedJsonObjectWithArrayAndWithConfig

hofi1 opened the pull request:
## Problem:
<!-- Explain the context and why you're making that change. What is the problem you're trying to solve? In some cases there is not a problem and this can be thought of being the motivation for your change. -->
In this test, the contents of an XML file and a JSON file is getting parsed and compared. But XML does not specify the order of the elements in the String, what means that the actual and the expected string cannot be directly compared for equality.
The flaky test was found by using the [NonDex](https://mvnrepository.com/artifact/edu.illinois/nondex-maven-plugin) tool.

https://github.com/hofi1/JSON-java/blob/01727fd0ed00c1f1f5582bdc97244356ac029f3b/src/test/java/org/json/junit/XMLTest.java#L1242

## Solution:
<!--Changed the assert statement from a JUnit Assertion to an XML Assertion and wrapped it in a default root element. This default root element is necessary, because of the specification of well-formed XML files. -->
To fix this assertion without the import of a new XMLAssertion library, just like in other PRs discussed, the XML String gets converted to a JSON Object and afterward compared for similarity. This is needed due to the fact that XML Objects do not offer a `similarity()` method.  

https://github.com/hofi1/JSON-java/blob/f4f7fc4a80f0999ec0b4957a7b7cc87fd49dcfa7/src/test/java/org/json/junit/XMLTest.java#L1242

## Result:
<!-- What will change as a result of your pull request? Note that sometimes this section is unnecessary because it is self-explanatory based on the solution. -->
The test is deterministic and not flaky. This improves the quality of the test and reduces the time to search for the bug during future development, as well as the need for reruns of the pipeline. 

## Reproduce
```shell
mvn edu.illinois:nondex-maven-plugin:2.1.1:nondex -Dtest=org.json.junit.XMLTest#testIndentComplicatedJsonObjectWithArrayAndWithConfig
```

eamonnmcmanus committed:
Small test fixes.

One test method was missing `@Test` so it was never run.

One test method used another test class as the base for finding a test
resource. While this works in practice with Maven, it is not strictly
right.

eamonnmcmanus committed:
Generalize the logic to disallow nested objects and arrays as keys in objects.

Fixes #771.

eedijs committed:
Add optJSONArray method to JSONObject with a default value

stleary committed:
Merge pull request #770 from eamonnmcmanus/testfixes

Small test fixes.

eedijs committed:
Add optJSONArray and optJSONObject methods to JSONArray with a default value

eamonnmcmanus committed:
Simplify the check for object keys that are themselves objects.

For object keys, we can just skip the part of `nextValue()` that parses values
that are objects or arrays. Then the existing logic for unquoted values will
already stop at `{` or `[`, and that will produce a `Missing value` exception.

eamonnmcmanus committed:
Add more test cases for unquoted text in objects and arrays.

eamonnmcmanus committed:
Merge branch 'stleary:master' into complexkey

KiruthikaJanakiraman commented:
The fix makes sense, but can you try to make a fix without adding a dependency to pom as adding a new dependency may cause security implications which might lead to rejection of the PR.

stleary committed:
Merge pull request #773 from eedijs/master

Add optJSONArray method to JSONObject with a default value

mccartney committed:
Removing excessive synchronization

mccartney committed:
Junit 4.13.2

stleary committed:
Merge pull request #772 from eamonnmcmanus/complexkey

Disallow nested objects and arrays as keys in objects.

Madjosz committed:
fix failing test XML test on Windows machines

Madjosz committed:
add validity check for JSONObject constructors

* fixes #713
* document JSONException in JavaDoc
* remove unused Comparable<T> boundary to reuse GenericBean in test

stleary committed:
Merge pull request #776 from mccartney/junit-4-13-2

JUnit 4.13.2

stleary committed:
Merge pull request #774 from mccartney/removing-synchronized

Removing excessive synchronization

hofi1 commented:
@KiruthikaJanakiraman thank you for your great feedback. 
I totally see your point and I tried to do it without the XMLAssert library, but since the XML strings which I am comparing are so big, it is not really feasible to write my own JUnit assertions for that as well as they are only available as strings, what makes it really hard to convert them in any other data structure (like a map or similar thing) to compare them.

mureinik committed:
Fix XMLTest on Windows

XMLTest.testIndentComplicatedJsonObjectWithArrayAndWithConfig fails
when run on Windows due to mismatching linebreaks (that aren't
important for the test's functionality) between the actual and
expected strings.

For the actual strings, linebreaks are canonized to the platform's
native linebreak using `replaceAll("\\n|\\r\\n",
System.getProperty("line.separator")`. However, the expected result is
read from a file, and is left with the linebreaks that were originally
used to create it.

The solution is to perform the same canonization on both strings.

Closes #781

mureinik committed:
Use System.lineSeparator()

Use the built-in System.lineSeparator() instead of implementing it
ourselves with System.getProperty("line.separator") in order to clean
up the code and make it easier to maintain.

rudrajyotib committed:
#653 - optLong vs getLong inconsistencies
For exponential decimal conversion, number is not touched.
Leading zeros removed from numeric number strings before converting to number.

stleary committed:
Merge pull request #782 from mureinik/XMLTest-windows

Fix XMLTest.testIndentComplicatedJsonObjectWithArrayAndWithConfig() for Windows - in the test

stleary committed:
Merge branch 'master' into fix_xml_test

stleary committed:
Merge pull request #778 from Madjosz/fix_xml_test

Fix XMLTest.testIndentComplicatedJsonObjectWithArrayAndWithConfig() for Windows - in the test

stleary committed:
Merge pull request #779 from Madjosz/713_jsonobject_nonfinite

add validity check for JSONObject constructors

rudrajyotib committed:
Merge branch 'stleary:master' into master

rudrajyotib committed:
#653 - review comments updated.

rudrajyotib committed:
#653 - review comments updated.

stleary committed:
Merge pull request #783 from rudrajyotib/master

optLong vs getLong inconsistencies

Commit:
update the docs for release 20231013

stleary committed:
Merge pull request #792 from stleary/pre-release-20231013

update the docs for release 20231013

Commit:
Reverting #761

stleary committed:
Merge pull request #793 from stleary/revert-761

Reverting #761

MyEnthusiastic commented:
Awesome work Simon!  🎉 congratulation for getting merged
